kernel/src/schema.rs

```diff
@@ -140,6 +143,38 @@ impl StructType {
         }
     }
 
+    /// Get a [`StructType`] containing [`StructField`]s of the given names, preserving the original
+    /// order of fields. Returns an Err if a specified field doesn't exist
+    pub fn project(&self, names: &[impl AsRef<str>]) -> DeltaResult<StructType> {
+        let mut indexes: Vec<usize> = names
+            .iter()
+            .map(|name| {
+                self.fields
+                    .get_index_of(name.as_ref())
```

Comment:
This will have `O(n*m)` cost, where `n` is the schema arity and `m` is the projection arity. Perhaps we could borrow spark's approach, which creates a set from the `names` (presumably `m` is much smaller than `n`) and then does a single filter-map pass over the fields, returning only those fields whose name is present in the set of names. Also avoids the other two passes (sort and index).